File: go/parsers/unregistered.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Unregistered struct {
	gen *gen.Unregistered
}

func NewUnregisteredFields(g *gen.Unregistered) messages.UnregisteredFields {
	return &Unregistered{gen: g}
}

func (u *Unregistered) RequestID() uint64 {
	return u.gen.RequestID()
}

func UnregisteredToCapnproto(m *messages.Unregistered) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	unregistered, err := gen.NewRootUnregistered(seg)
	if err != nil {
		return nil, err
	}

	unregistered.SetRequestID(m.RequestID())

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeUnregistered, data, nil), nil
}

func CapnprotoToUnregistered(data []byte) (*messages.Unregistered, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	ur, err := gen.ReadRootUnregistered(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewUnregisteredWithFields(NewUnregisteredFields(&ur)), nil
}

File: python/wamp_msgs_capnp/parsers/unregistered.py
import os
from pathlib import Path

import capnp
from wampproto.messages.unregistered import Unregistered, IUnregisteredFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
schema_file = os.path.join(root_dir, "schemas", "unregistered.capnp")
unregistered_capnp = capnp.load(str(schema_file))


class UnregisteredFields(IUnregisteredFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def request_id(self) -> int:
        return self._gen.requestID


def unregistered_to_capnproto(u: Unregistered) -> bytes:
    unregistered = unregistered_capnp.Unregistered.new_message()
    unregistered.requestID = u.request_id

    packed_data = unregistered.to_bytes_packed()

    return helpers.prepend_header(Unregistered.TYPE, packed_data, b"")


def capnproto_to_unregistered(data: bytes) -> Unregistered:
    message_data, _ = helpers.extract_message(data)
    unregistered_obj = unregistered_capnp.Unregistered.from_bytes_packed(message_data)

    return Unregistered(UnregisteredFields(unregistered_obj))
